### PR TITLE
Lev reworked locations

### PR DESCRIPTION
**Pull request by levb**
- [ ] implement channel header button

### CODE REVIEW

File: server/api/app.go [hunk 1]
```diff
@@ -7,15 +7,18 @@ import (
...
+	OAuth2CallbackURL string `json:"oauth2_callback_url,omitempty"`
+	HomepageURL       string `json:"homepage_url,omitempty"`
+	RootURL           string `json:"root_url"`
+
 	RequestedPermissions Permissions `json:"requested_permissions,omitempty"`
-	RootURL              string      `json:"root_url"`
-	LocationsURL         string      `json:"location_url"`
+
+	Install   string `json:"install,omitempty"`
+	Functions string `json:"functions"`
```
Comment (levb) [inline server/api/app.go]: []string

File: server/apps/call.go [hunk 2]
```diff
@@ -3,15 +3,19 @@ package apps
 import "github.com/mattermost/mattermost-plugin-apps/server/api"
 
 func (s *service) Call(call *api.Call) (*api.CallResponse, error) {
-	var err error
+	// TODO <><> get a cached function
```
Comment (levb) [inline server/apps/call.go:6]: ticket: Implement the function meta cache

File: server/apps/client.go [hunk 3]
```diff
@@ -23,9 +24,9 @@ const OutgoingAuthHeader = "Mattermost-App-Authorization"
 
 type Client interface {
 	GetManifest(manifestURL string) (*api.Manifest, error)
-	PostCall(call *api.Call) (*api.CallResponse, error)
+	PostFunction(call *api.Call) (*api.CallResponse, error)
 	PostNotification(n *api.Notification) error
-	GetLocations(appID api.AppID, userID, channelID string) ([]api.LocationInt, error)
+	GetFunction(call *api.Call) (*api.Function, error)
```
Comment (levb) [inline server/apps/client.go]: GetFunctionMeta

File: server/apps/install.go [hunk 4]
```diff
@@ -44,7 +45,7 @@ func (s *service) InstallApp(in *api.InInstallApp, cc *api.Context, sessionToken
 		return nil, "", err
 	}
 
-	ExpandedContext, err := s.newExpander(cc).Expand(
+	exp, err := s.newExpander(cc).Expand(
```
Comment (levb) [inline server/apps/install.go]: ecc

File: server/apps/locations.go [hunk 5]
```diff
@@ -1,26 +1,10 @@
...
-		locations, err := s.Client.GetLocations(id, userID, channelID)
-		if err != nil {
-			return nil, errors.Wrap(err, "failed to get single location")
-		}
-		allLocations = append(allLocations, locations...)
-	}
-
-	return allLocations, nil
+func (s *service) GetBindings(cc *api.Context) ([]*api.Binding, error) {
+	return nil, nil
```
Comment (levb) [inline server/apps/locations.go]: Implement

File: server/command/main.go [hunk 6]
```diff
@@ -57,9 +57,10 @@ func (s *service) executeDebugClean(params *params) (*model.CommandResponse, err
 }
 
 func (s *service) executeDebugLocations(params *params) (*model.CommandResponse, error) {
-	locations, err := s.apps.API.GetLocations(params.commandArgs.UserId, params.commandArgs.ChannelId)
-	if err != nil {
-		return normalOut(params, md.MD("error"), err)
-	}
-	return normalOut(params, md.JSONBlock(locations), nil)
+	// locations, err := s.apps.API.GetLocations(params.commandArgs.UserId, params.commandArgs.ChannelId)
```
Comment (levb) [inline server/command/main.go]: restore

File: server/http/restapi/locations.go [hunk 7]
```diff
@@ -4,32 +4,31 @@ import (
...
-	"github.com/pkg/errors"
 )
 
 func (a *restapi) handleLocations(w http.ResponseWriter, req *http.Request, actingUserID string) {
-	userID := req.URL.Query().Get("user_id")
-	if userID == "" {
-		httputils.WriteBadRequestError(w, errors.New("no user id"))
-		return
-	}
+	// userID := req.URL.Query().Get("user_id")
```
Comment (levb) [inline server/http/restapi/locations.go]: Implement

File: server/api/binding.go [hunk 8]
```diff
@@ -0,0 +1,45 @@
...
+
+type LocationID string
+
+const (
+	LocationPostMenu      LocationID = "/post_menu"
+	LocationChannelHeader LocationID = "/channel_header"
+	LocationCommand       LocationID = "/command"
+)
+
+type Binding struct {
```
Comment (larkox) [inline server/api/binding.go:11]: I find this not enough for expanding locations. e.g. In mobile some channel header menu items have toggle status or a count. If we wanted to implement that, we could not codify that with this structure.

File: server/api/call.go [hunk 9]
```diff
@@ -8,12 +8,11 @@ import (
 )
 
 type Call struct {
-	FormURL string      `json:"form_url,omitempty"`
-	Values  FormValues  `json:"values,omitempty"`
-	Context *Context    `json:"context,omitempty"`
-	Expand  *Expand     `json:"expand,omitempty"`
-	AsModal bool        `json:"as_modal,omitempty"`
-	From    []*Location `json:"from,omitempty"`
```
Comment (larkox) [inline server/api/call.go:16]: Are we no longer interested on the From field? or has it been moved?

**Review by larkox — COMMENTED** [review]
In general it seems to be cleaner, but I am worried about two main things:
- Locations becoming hard to expand (adding new functionality, or new UI elements...)
- Being too strict with the naming (i.e. locations must be stored in rootURL/constants.BindingsPath, instead of wherever the app wants).

**Pull request closed without merging**